mention: Fetch users to be notified of @topic mentions.

prakhar1144 opened the pull request:
[#25828 (comment)](https://github.com/zulip/zulip/pull/25828#pullrequestreview-1514442343)
> I think the group of first 6 commits would be a good prefix to split into a new PR so that it can be merged off. The notifications commit is pretty large and would be easier to review independently :)

amanagr commented:
@abhijeetbodas2001 I guess you will review this first?

abhijeetbodas2001 commented:
I had received the commits here as part of #25828. This looks good to me now (Prakhar has addressed my comments from there).

prakhar1144 commented:
@zulipbot add "integration review"

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -217,12 +225,22 @@ def get_recipient_info(
+            # A topic participant is anyone who either sent or reacted to messages in the topic.
+            # It is expensive to call `participants_for_topic` if the topic has a large number
+            # of messages. But it is fine to call it here, as this gets called only if the message
+            # has @topic mentions.
```
Most correctly, we only call this when there's syntax that might be an `@topic` mention, without having confirmed the syntax isn't, say, in a code block.

prakhar1144 replied:
Updated comment:
```python
  # A topic participant is anyone who either sent or reacted to messages in the topic.
  # It is expensive to call `participants_for_topic` if the topic has a large number
  # of messages. But it is fine to call it here, as this gets called only if the message
  # has syntax that might be a @topic mention without having confirmed the syntax isn't, say,
  # in a code block.
```

timabbott commented on zerver/tests/test_markdown.py:
```diff
@@ -1813,6 +1814,20 @@ def test_default_code_block_language(self) -> None:
+            rendering_result = render_markdown(msg, content)
+            self.assertEqual(
+                rendering_result.rendered_content,
+                f'<p><span class="user-mention" data-user-id="*">@{topic_wildcard}</span> test</p>',
```
Hmm, can you start an "api design" conversation about this `data-user-id` value? I think it's possible we want a special value of "topic" or some other indication that is the a topic mention; I guess simplest would be to keep the `data-user-id` as `*` but also have a `topic-mention` class on the `user-mention` element. As it is, the only way for clients to determine whether a given mention inside a message was a topic mention is by parsing the text inside, and we definitely don't want to use that mechanism, since it's not at all robust for future changes to what syntax we use.

timabbott commented:
This overall looks pretty good; I think https://github.com/zulip/zulip/pull/26195#discussion_r1254997553 needs to be resolved before merging, but otherwise I think this PR is ready. Thanks for breaking this out as a separate mini PR!

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments (Please check the respective review comments)
* Pending [CZO discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/.40topic-mention.20data-user-id/near/1605762)

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1812,13 +1818,16 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
                 el.set("class", "user-mention")
                 text = f"@{text}"
+            if topic_wildcard:
+                el.set("class", el.get("class", "") + " topic-mention")
```
I'm not sure we want `user-mention` as well in this case; we can discuss here:

https://chat.zulip.org/#narrow/stream/378-api-design/topic/.40topic-mention.20data-user-id.20.2326195/near/1608185

prakhar1144 replied:
Removed `user-mention` after discussion in CZO

timabbott commented:
Merged, thanks @prakhar1144!

Please open issues with the mobile and terminal repositories for rendering topic mentions correctly when displayed in a message feed.
